### PR TITLE
Notifations

### PR DESCRIPTION
**Pull request by xx-TOOL-xx**
Created different types of notifications, added some features. Now user can't send proposa/invite/offer for same job to same freelancer. Changed some messages functionality

UPD: added redis to improve sockets functionality

### CODE REVIEW

File: src/modules/chat-room/chat-room.service.ts [hunk 1]
```diff
@@ -94,6 +94,31 @@ export class ChatRoomService {
       .orderBy('updated_at', 'DESC');
   }
 
+  async getRoomIdByJobPostAndFreelancerIds(
+    jobPostId: number,
+    freelancerId: number,
+  ) {
```
Comment (milkavladislav) [inline src/modules/chat-room/chat-room.service.ts]: return value type `Promise<number>`

File: src/modules/chat-room/chat-room.service.ts [hunk 2]
```diff
@@ -94,6 +94,31 @@ export class ChatRoomService {
...
+      .leftJoin('chat_room.jobPost', 'jobPost')
+      .leftJoin('chat_room.freelancer', 'freelancer')
+      .where(`jobPost.id = ${jobPostId}`)
+      .andWhere(`freelancer.id = ${freelancerId}`)
+      .getOneOrFail();
+
+    return room.id;
+  }
+
+  async getRoomIdByMessageId(id: number) {
```
Comment (milkavladislav) [inline src/modules/chat-room/chat-room.service.ts]: here also

File: src/modules/message/message.service.ts [hunk 3]
```diff
@@ -19,13 +27,13 @@ export class MessageService {
     @InjectRepository(ChatRoom)
     private readonly chatRoomRepository: Repository<ChatRoom>,
```
Comment (milkavladislav) [inline src/modules/message/message.service.ts]: unnecessary repository

File: src/modules/message/message.service.ts [hunk 4]
```diff
@@ -48,19 +56,73 @@ export class MessageService {
...
+    chatRoomId: number,
+    inviteFrom: number,
+    inviteTo: number,
+    type: ProposalType,
+    message: string,
+    bid: number,
+  ): Promise<void> {
+    await this.createSystemMessage({
+      text: `You have received a new ${
+        type === ProposalType.PROPOSAL ? 'proposal' : 'invite'
```
Comment (milkavladislav) [inline src/modules/message/message.service.ts]: maybe you should just let "`You have received a new ${type}`"

Comment (milkavladislav) [reply src/modules/message/message.service.ts:89]: maybe you should just let `You have received a new ${type}`.

File: src/modules/notifications/notifications.service.ts [hunk 5]
```diff
@@ -1,45 +1,230 @@
...
+  ICountedParsedNotifications,
+} from 'src/modules/notifications/typesDef';
 
 @Injectable()
 export class NotificationsService {
   constructor(
     @InjectRepository(Notification)
     private notificationRepository: Repository<Notification>,
+    @Inject(forwardRef(() => MessageService))
+    private messageService: MessageService,
```
Comment (milkavladislav) [inline src/modules/notifications/notifications.service.ts]: unnecessary service

File: src/modules/notifications/notifications.service.ts [hunk 6]
```diff
@@ -1,45 +1,230 @@
...
+    userId: number,
+    type: ProposalType,
   ): Promise<Notification> {
-    const newNotification = await this.notificationRepository.save(
-      createNotificationDto,
-    );
-    return this.notificationRepository
+    return await this.create({
+      type: NotificationType.PROPOSAL,
+      text: `You have receive a new ${
```
Comment (milkavladislav) [inline src/modules/notifications/notifications.service.ts]: same, maybe you should just let "`You have received a new ${type}`"

File: src/modules/proposal/proposal.service.ts [hunk 7]
```diff
@@ -21,68 +23,99 @@ export class ProposalService {
     private readonly chatRoomService: ChatRoomService,
     private readonly freelancerService: FreelancerService,
     private readonly settingsInfoService: SettingsInfoService,
```
Comment (milkavladislav) [inline src/modules/proposal/proposal.service.ts]: unnecessary service

File: src/modules/proposal/proposal.service.ts [hunk 8]
```diff
@@ -21,68 +23,99 @@ export class ProposalService {
...
-    @InjectRepository(Message)
-    private readonly messageRepo: Repository<Message>,
+    private readonly messageService: MessageService,
+    private readonly jobPostService: JobPostService,
+    @InjectRepository(LocalFile)
+    private readonly local: Repository<LocalFile>,
```
Comment (milkavladislav) [inline src/modules/proposal/proposal.service.ts]: also unnecessary

File: src/modules/proposal/proposal.service.ts [hunk 9]
```diff
@@ -21,68 +23,99 @@ export class ProposalService {
...
+    private readonly notificationsService: NotificationsService,
   ) {}
 
   async create(
     createProposalDto: CreateProposalDto,
     userId: number,
```
Comment (milkavladislav) [inline src/modules/proposal/proposal.service.ts:44]: user unnecessary here

**Review by milkavladislav — APPROVED** [review]

File: src/modules/chat-room/chat-room.service.ts (hunk 1 above)
Comment (xx-TOOL-xx) [reply src/modules/chat-room/chat-room.service.ts]: fixed

File: src/modules/chat-room/chat-room.service.ts (hunk 2 above)
Comment (xx-TOOL-xx) [reply src/modules/chat-room/chat-room.service.ts]: fixed

File: src/modules/message/message.service.ts (hunk 3 above)
Comment (xx-TOOL-xx) [reply src/modules/message/message.service.ts]: fixed

File: src/modules/message/message.service.ts (hunk 4 above)
Comment (xx-TOOL-xx) [reply src/modules/message/message.service.ts]: I don't think so. Values of enum uses for functionality, and they can be renamed or be in camel cace etc, so I thinck it better use static text, it's just a coincidence that their texts are same

File: src/modules/notifications/notifications.service.ts (hunk 5 above)
Comment (xx-TOOL-xx) [reply src/modules/notifications/notifications.service.ts]: fixed

File: src/modules/notifications/notifications.service.ts (hunk 6 above)
Comment (xx-TOOL-xx) [reply src/modules/notifications/notifications.service.ts]: same

File: src/modules/proposal/proposal.service.ts (hunk 7 above)
Comment (xx-TOOL-xx) [reply src/modules/proposal/proposal.service.ts]: fixed

File: src/modules/proposal/proposal.service.ts (hunk 8 above)
Comment (xx-TOOL-xx) [reply src/modules/proposal/proposal.service.ts]: fixed

File: src/modules/message/message.service.ts [hunk 10]
```diff
@@ -16,76 +26,161 @@ export class MessageService {
...
+    inviteFrom: number,
+    inviteTo: number,
+    type: ProposalType,
+    message: string,
+    bid: number,
+  ): Promise<void> {
+    try {
+      await this.createSystemMessage({
+        text: `You have received a new ${
+          type === ProposalType.PROPOSAL ? 'proposal' : 'invite'
```
Comment (lyudad) [inline src/modules/message/message.service.ts]: dont use string in the code, move to constants

File: src/modules/notifications/notifications.service.ts [hunk 11]
```diff
@@ -1,45 +1,335 @@
...
+    userId: number,
+    type: ProposalType,
   ): Promise<Notification> {
-    const newNotification = await this.notificationRepository.save(
-      createNotificationDto,
+    try {
+      return await this.create({
+        type: NotificationType.PROPOSAL,
+        text: `You have receive a new ${
+          type === ProposalType.PROPOSAL ? 'proposal' : 'invite'
```
Comment (lyudad) [inline src/modules/notifications/notifications.service.ts]: same

**Pull request merged**